sycl/doc/extensions/proposed/sycl_ext_intel_fpga_pipes_properties.asciidoc

```diff
@@ -0,0 +1,264 @@
+`sycl_ext_oneapi_properties` extensions.
+
+== Overview
+
+This extension introduces properties that establish differences in the
+implementation of `sycl::pipe`. These properties are FPGA specific. An example
+of the syntax can be seen below. 
+
+[source,c++]
+----
+using pipe = pipe<class some_pipe, int, decltype(properties{uses_ready<true>})>;
+----
+
+== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification section 6.3.3 "Feature test macros". Therefore, an implementation
+supporting this extension must predefine the macro
+`SYCL_EXT_INTEL_FPGA_PIPE_PROPERTIES` to one of the values defined in the table
+below. Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's APIs the implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value |Description
+|1     |Initial extension version.  Base features are supported.
+|===
+
+=== Pipe properties
+
+Below is a list of compile-time-constant properties which `pipe` supports.
+
+```c++
+namespace sycl::ext::intel::experimental {
+
+struct min_capacity {
+  template<int Capacity>
+  using value_t = property_value<min_capacity, std::integral_constant<int, Capacity>>;
+};
+
+struct ready_latency {
+  template<int Latency>
+  using value_t = property_value<ready_latency, std::integral_constant<int, Latency>>;
+};
+
+struct bits_per_symbol {
+  template<int Bits>
+  using value_t = property_value<bits_per_symbol, std::integral_constant<int, Bits>>;
+};
+
+struct uses_valid {
+  template<bool Valid>
+  using value_t = property_value<uses_valid, std::integral_constant<bool, Valid>>;
+};
+
+struct uses_ready {
+  template<bool Ready>
+  using value_t = property_value<uses_ready, std::integral_constant<bool, Ready>>;
+};
+
+struct first_symbol_in_high_order_bits {
+  template<bool HighOrder>
+  using value_t = property_value<first_symbol_in_high_order_bits, std::integral_constant<bool, HighOrder>>;
+};
+
+struct protocol {
+  enum class protocol_name {
+    AVALON,
+    AXI
+  };
+
+  template<protocol_name Protocol>
+  using value_t = property_value<protocol, std::integral_constant<protocol_name, Protocol>>;
+};
+
+} // namespace sycl::ext::intel::experimental
+```
+
+--
+[options="header"]
+|====
+| Property | Description
+|`min_capacity`
+| Valid values: Non-negative integer value.
+
+Default value: 0
+
+User defined minimum number of words in units of data type size that the pipe
+must be able to store without any being read out. A minimum capacity is required
+in some algorithms to avoid deadlock, or for performance tuning. An
+implementation can include more capacity than this parameter, but not less.
+
+This property is not guaranteed to be respected if the pipe is an inter-kernel
+pipe. The compiler is allowed to optimize the pipe if both sides are visible.
```

Comment:
I don't understand how this can be ignored for inter-kernel pipes.  As you say, the algorithm may depend on a certain minimum capacity.

In any event, isn't it always assumed that a compiler can optimize away code if it can prove that doing so has no effect?  I don't think we need to say that in the spec if that's all we mean.